service/history/transferQueueProcessor.go

```diff
@@ -584,6 +590,105 @@ func (t *transferQueueProcessorImpl) processCancelExecution(task *persistence.Tr
 	return err
 }
 
+func (t *transferQueueProcessorImpl) processSignalExecution(task *persistence.TransferTaskInfo) error {
+	t.metricsClient.IncCounter(metrics.TransferTaskSignalExecutionScope, metrics.TaskRequests)
+	sw := t.metricsClient.StartTimer(metrics.TransferTaskSignalExecutionScope, metrics.TaskLatency)
+	defer sw.Stop()
+
+	var err error
+	domainID := task.DomainID
+	targetDomainID := task.TargetDomainID
+	execution := workflow.WorkflowExecution{WorkflowId: common.StringPtr(task.WorkflowID),
+		RunId: common.StringPtr(task.RunID)}
+
+	var context *workflowExecutionContext
+	var release releaseWorkflowExecutionFunc
+	context, release, err = t.cache.getOrCreateWorkflowExecution(domainID, execution)
+	defer release()
+	if err != nil {
+		return err
+	}
+
+	var msBuilder *mutableStateBuilder
+	msBuilder, err = context.loadWorkflowExecution()
+	if err != nil {
+		if _, ok := err.(*workflow.EntityNotExistsError); ok {
+			// this could happen if this is a duplicate processing of the task, and the execution has already completed.
+			return nil
+		}
+		return err
+	}
+
+	initiatedEventID := task.ScheduleID
+	ri, isRunning := msBuilder.GetSignalInfo(initiatedEventID)
+	if !isRunning {
+		return nil
+	}
+
+	signalRequest := &history.SignalWorkflowExecutionRequest{
+		DomainUUID: common.StringPtr(targetDomainID),
+		SignalRequest: &workflow.SignalWorkflowExecutionRequest{
+			Domain: common.StringPtr(targetDomainID),
+			WorkflowExecution: &workflow.WorkflowExecution{
+				WorkflowId: common.StringPtr(task.TargetWorkflowID),
+				RunId:      common.StringPtr(task.TargetRunID),
+			},
+			Identity:   common.StringPtr(identityHistoryService),
+			SignalName: common.StringPtr(ri.SignalName),
+			Input:      ri.Input,
+			// Use same request ID to dedupe SignalWorkflowExecution calls
+			RequestId: common.StringPtr(ri.SignalRequestID),
+		},
+	}
+
+	op := func() error {
```

Comment:
Can you just create a helper for SignalExecutionWithRetry.

Reply:
changed